ksql-serde/src/main/java/io/confluent/ksql/serde/json/JsonSerdeUtils.java

```diff
@@ -23,14 +23,59 @@
 import io.confluent.ksql.schema.connect.SchemaWalker.Visitor;
 import io.confluent.ksql.schema.ksql.PersistenceSchema;
 import io.confluent.ksql.schema.ksql.SqlBaseType;
+import io.confluent.ksql.util.KsqlException;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import javax.annotation.Nonnull;
 import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.connect.data.Schema.Type;
 
-final class JsonSerdeUtils {
+public final class JsonSerdeUtils {
+
+  // the JsonSchemaConverter adds a magic NULL byte and 4 bytes for the
+  // schema ID at the start of the message
+  private static final int SIZE_OF_SR_PREFIX = Byte.BYTES + Integer.BYTES;
 
   private JsonSerdeUtils() {
   }
 
+  /**
+   * Converts {@code jsonWithMagic} into an {@link InputStream} that represents
+   * standard JSON encoding.
+   *
+   * @param jsonWithMagic the serialized JSON
+   * @return the corresponding input stream
+   * @throws io.confluent.ksql.util.KsqlException If the input is not encoded
+   *         using the schema registry format (first byte magic byte, then
+   *         four bytes for the schemaID).
+   */
+  public static InputStream asStandardJson(@Nonnull final byte[] jsonWithMagic) {
```

Comment:
Why does this need to return an InputStream? It seems that all the callers really just want the bytes.

Would it be simpler (and more efficient) just to slice the byte[] and return another one with the prefix removed?

Also.. I know this InputStream won't block, but as we move to a more reactive / non blocking model in the server we should avoid using blocking constructs such as input and output streams and favour buffers.

Reply:
I think Almog is trying to avoid the array copy for every message when deserializing. This could be premature optimisation, but I'm guessing it will hurt performance as we already know serialization is a large cost to us.

An alternative to copying the data into a new buffer would be to build a parser with the original buffer _and_ with the suitable offset:, i.e. change a line like:

```java
// Original code that deserialized the whole byte array:
MAPPER.readTree(bytes);
```

to

```java
// Build a parser with the whole array and appropriate offset into that array:
// This avoids the array copy at the cost of slightly more complex code.
final int offset = isJsonSchema ? JsonSerdeUtils.SIZE_OF_SR_PREFIX : 0;
MAPPER.readTree(MAPPER.getFactory().createParser(bytes, offset, bytes.length - offset));
```

Reply:
Copying small arrays is super fast in Java.

Reply:
Also if you use ByteBuffer or Vert.x Buffers you can avoid the copy altogether as slice just references the original array with different offset and length.

Reply:
replaced with:
```java
    return mapper.readValue(
        jsonWithMagic,
        SIZE_OF_SR_PREFIX,
        jsonWithMagic.length - SIZE_OF_SR_PREFIX,
        clazz
    );
```
Inside the method. I believe this should work.